### DISCUSSION TITLE
Adding discrete units to new capacity

### DISCUSSION THREAD

**Discussion by Alaa-Eltablawy** [iiasa/message_ix#838] [MESSAGE modeling]
Does MessageIX supports adding discrete units of a technology to new capacity?

For example, if we have a demand of 150 KW and the technology we have is solar panels where the max capacity for the panel is 100KW.
In this case, will messageIX assumes that we will install 1.5 solar panels? or is there a way where we have integer increment for this technology, so that we will have 2 solar panels installed and the investment cost will be for 2 panels?

**Comment by glatterf42** [comment]
Hi @Alaa-Eltablawy, thanks for this question! 
In [our documentation](https://docs.messageix.org/en/latest/model/MESSAGE/model_core.html#decision-variables), we specify that all capacity-related decision variables (such as `CAP_NEW`, the newly installed capacity) are real numbers above (or including) 0, so in theory, any value can be built in the model. There are several options to constrain this: you could use [fixed boundaries](https://docs.messageix.org/en/latest/model/MESSAGE/parameter_def.html#bounds-on-capacity-and-activity) or [dynamic boundaries](https://docs.messageix.org/en/latest/model/MESSAGE/parameter_def.html#dynamic-constraints-on-new-capacity-and-activity) to constrain the minimum and maximum of newly or total built capacity. 
What I haven't found from our docs is how you would limit a technology such that one installment of it has a maximum capacity of e.g. 100kW. From our [list of Parameters for technologies](https://docs.messageix.org/en/latest/model/MESSAGE/parameter_def.html), `output` sounds closest to it, i.e. you would limit the `output` of a `solar panel` technology to 100kW. Is this indeed what you do?
In that case, I think the first part of this answer is what you are looking for: the model can built any percentage of a technology as `CAP_NEW`, so it would probably build 1.5 `solar panel`s in your case.

**Comment by Alaa-Eltablawy** [comment]
Thanks @glatterf42 for your reply!
I believe the 2nd half of your answer is quite close to what I want to implement.
Currently messageIX supports setting bound_total_capacity_up, so as you mentioned in your answer, CAP_NEW will be a real number.
While in some cases, the technology itself might be in form of units, so the CAP_NEW should be always integer value for this kind of unit based technology, since it won't be realistic to install only half of solar panel, or half of a wind turbine.
I believe in that case, we will need to set the bound capacity up for the single unit of the technology, does "output" supports that?
Also I believe having support for the unit based technologies will change the problem from linear programming to mixed integer linear programming problem.

**Reply by SiddharthJoshi-Git** [reply]
Dear @Alaa-Eltablawy,

I will share my experience here, in addition to @glatterf42 .

To start with, `demand` in our model is represented in terms of generation requirements i.e. kWh/GWa etc. The conversion from demand (generation) to capacity is done using Capacity Factors (C.F.s).

It is not possible to force MESSAGEix to provide integer solutions e.g. output 2GW capacity instead of 1.5GW capacity. The core objective in the optimisation is to match supply-demand with least cost. This would mean that real numbers are allowed for demand, supply, generation and capacity. The investment costs are assigned as an integer e.g. 1000$/kW, here the final investment amount will become a real number based on fractional capacity addition. 

Changing `output` parameter is not recommended as input-output parameters are generally used for defining technology `efficiencies`, technology` input-output` flows. https://docs.messageix.org/en/latest/efficiency.html

Adding `bounds` on deployments (Capacity) can only act as bounding constraints if demand-supply is asking for it.

So, in summary we don't run MESSAGEix in an `integer LP/ Mix integer LP` mode which will ensure the result that you are seeking. This is by design as the scenarios we develop are generally in the scale of '1000 GW solar capacity, where we do not need the precision of kW scale.
